x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/CopySign.java

```diff
@@ -0,0 +1,159 @@
+    ) {
+        super(source, Arrays.asList(magnitude, sign));
+    }
+
+    private CopySign(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(children().get(0));
+        out.writeNamedWriteable(children().get(1));
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, CopySign::new, children().get(0), children().get(1));
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        if (newChildren.size() != 2) {
+            throw new EsqlIllegalArgumentException("Function [{}] expects exactly two arguments, got [{}]", NAME, newChildren.size());
+        }
+        return new CopySign(source(), newChildren.get(0), newChildren.get(1));
+    }
+
+    @Override
+    public DataType dataType() {
+
+        if (dataType == null) {
+            resolveType();
+        }
+        return dataType;
+    }
+
+    @Override
+    public TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        var magnitude = children().get(0);
+        var sign = children().get(1);
+        if (magnitude.dataType().isNumeric() == false) {
+            return new TypeResolution("Magnitude must be a numeric type");
+        }
+        if (sign.dataType().isNumeric() == false) {
+            return new TypeResolution("Sign must be a numeric type");
+        }
+        // TODO(pabloem): Looks like we don't currently support generalizing types
+        // when two types are incompatible (e.g. copySign(int, long) -> long instead of
+        // copySign(int, long) => ERROR).
+        dataType = magnitude.dataType();
+        TypeResolution resolution = TypeResolutions.isType(
+            magnitude,
+            t -> t == dataType,
+            sourceText(),
+            TypeResolutions.ParamOrdinal.fromIndex(1),
+            magnitude.dataType().typeName()
+        );
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
+        var dataType = dataType();
+        EvalOperator.ExpressionEvaluator.Factory magnitudeEvaluatorFactory = toEvaluator.apply(children().get(0));
+        EvalOperator.ExpressionEvaluator.Factory signEvaluatorFactory = toEvaluator.apply(children().get(1));
+        if (dataType == DataType.FLOAT) {
```

Comment:
The reason you had to modify the generated code for `FLOAT` is that we haven't been using that type yet. All `float` typed fields in the mappings are loaded as `DOUBLE` for historical reasons and we haven't made the breaking change to load them as `FLOAT`. So we haven't done much with `FLOAT` types. But it doesn't hurt anything to make it.